modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java

```diff
@@ -2644,78 +2651,128 @@ private synchronized void stop0(boolean cancel, boolean waitForBackups) {
-                            if (log != null) {
-                                LT.warn(log, "Delaying node shutdown due to ongoing topology change [curTopVer=" +
-                                    curTopVer + ", readyTopVer=" + readyTopVer + "]");
-                            }
+                    try {
+                        originalNodesToExclude = metaStorage.read(GRACEFUL_SHUTDOWN_METASTORE_KEY);
+
+                        nodesToExclude = originalNodesToExclude != null ? new HashSet<>(originalNodesToExclude) :
+                            new HashSet<>();
+                    }
+                    catch (IgniteCheckedException e) {
+                        U.error(log, "Unable to read " + GRACEFUL_SHUTDOWN_METASTORE_KEY +
+                            " value from metastore.", e);
 
-                            blocked = true;
+                        continue;
+                    }
 
-                            break;
-                        }
+                    for (CacheGroupContext grpCtx : grid.context().cache().cacheGroups()) {
+                        if (grpCtx.systemCache())
+                            continue;
 
-                        if (cctx.config().getCacheMode() == LOCAL)
+                        if (grpCtx.isLocal())
                             continue;
 
-                        int locNodeMinNumOfParts = cctx.group().metrics().getLocalNodeMinimumNumberOfPartitionCopies();
-                        if (locNodeMinNumOfParts == 1) {
-                            if (cctx.config().getCacheMode() == PARTITIONED && cctx.config().getBackups() == 0) {
-                                if (log != null) {
-                                    LT.warn(log, "Ignoring potential data loss on cache without backups [name=" +
-                                        cache.getName() + "]");
+                        if (grpCtx.config().getCacheMode() == PARTITIONED && grpCtx.config().getBackups() == 0) {
+                            U.warn(log, "Ignoring potential data loss on cache without backups [name="
+                                + grpCtx.cacheOrGroupName() + "]");
+
+                            continue;
+                        }
+
+                        GridDhtPartitionFullMap fullMap = grpCtx.topology().partitionMap(false);
+
+                        int cacheSpecificAmountOfOwners = -1;
+
+                        if (fullMap != null) {
+                            nodesToExclude.retainAll(fullMap.keySet());
+
+                            int parts = grpCtx.topology().partitions();
+
+                            for (int part = 0; part < parts; part++) {
+                                Set<UUID> partIdealAssignment = grpCtx.affinity().idealAssignmentRaw().get(part).
+                                    stream().map(ClusterNode::id).collect(Collectors.toSet());
+
+                                int cnt = 0;
+
+                                for (Map.Entry<UUID, GridDhtPartitionMap> entry : fullMap.entrySet()) {
```

Comment:
We basically execute the same algorithm as in GridDhtPartitionsExchangeFuture#updateMaximumAmountOfNodesAvailableForSafeStop. Can we extract it to one common place instead of copypasting?

Reply:
It's not exact the same, cause code inside `GridDhtPartitionsExchangeFuture#updateMaximumAmountOfNodesAvailableForSafeStop` doesn't contain metastore-stored-excluded-nodes logic. Seems that code modification that  will support such distinction will be a sort of overdesign.

Reply:
Reworked with 
`                        if (prevReadyTopVer == AffinityTopologyVersion.NONE.topologyVersion())
                            prevReadyTopVer = grpCtx.topology().readyTopologyVersion().topologyVersion();
                        else if (prevReadyTopVer != grpCtx.topology().readyTopologyVersion().topologyVersion()) {
                            // At the moment, there is an exchange.
                            safeToStop = false;

                            break;
                        }`